Task/FP-547: Relaunch jobs in Job History modal.

duckonomy opened the pull request:
## Overview: ##
Tried to replicate CEPv1's job relaunch behavior.

It seems like resubmit has only one api:
* In the Tapis documentation:
https://tapis-cli.readthedocs.io/en/latest/usage/jobs.html#jobs-resubmit
* Based on code from CEPv1:
https://bitbucket.org/taccaci/core-exp-portal/src/b633800261b4ef859a4bc36dacd5f8e5c0bc6897/server/portal/apps/workspace/api/views.py#lines-258
https://bitbucket.org/taccaci/core-exp-portal/src/b633800261b4ef859a4bc36dacd5f8e5c0bc6897/client/src/angular/workspace/services/jobs-service.js#lines-71

Based on CEPv1, we are assuming that job resubmission is done when job is finished.
* Based on code from CEPv1:
https://bitbucket.org/taccaci/core-exp-portal/src/b633800261b4ef859a4bc36dacd5f8e5c0bc6897/client/src/angular/jobs/components/job-actions/job-actions.template.html#lines-4

Design of the relaunch button was based on:
- https://xd.adobe.com/view/db2660cc-1011-4f26-5d31-019ce87c1fe8-ad17/screen/7d3e671f-bb61-4a20-89c6-9ca9a78a6f93/

## Related Jira tickets: ##

* [FP-547](https://jira.tacc.utexas.edu/browse/FP-547)

## Summary of Changes: ##
1. Added backend code the handle job relaunch by adding job objects as well as sending the job request to tapis.
2. Added the `Relaunch Job` button in the Job History modal.
3. Added chaining in the `SUBMIT_JOB` saga to call `GET_JOBS` after the resubmit job request.

## Testing Steps: ##
1. Submit a job that completes.
2. Once it's finished, ensure that you see the `Relaunch Job` button the Job's `View Details` modal (`workbench/history/jobs`).
3. Click `Relaunch Job`.
4. Ensure that a new task with the same parameters is launched and the previous job is removed from the list.
5. Ensure that the new task is successful.

## UI Photos:
![Screenshot from 2021-11-16 16-00-44](https://user-images.githubusercontent.com/9425579/142072938-45c8bcaf-ee9c-4615-a5c8-ac8c963ee336.png)

## Notes: ##
Questions:
1. Should we redesign it so that clicking on `Relaunch Job` opens another form that let's the user modify parameters based on the previous job submission?
2. Would it be correct to show the `Relaunch Job` button when job status is failed as well as finished?
3. Should we implement `cancelJob()` when job status is not finished? (code was already there in the backend)
4. I currently remove the previous job on clicking `Relaunch Job` and create a new job in it's place. Is this acceptable?
5. Should there be a separate `RESUBMITTING` job status?

TODO
1. [ ] Add test for dispatch events in job history modal.

xandroc reviewed (approved):
let's roll

nathanfranklin commented on server/portal/apps/workspace/api/views.py:
```diff
@@ -263,10 +263,28 @@ def post(self, request, *args, **kwargs):
+                if "id" in data:
+                    # Remove previous job fom job list
+                    prev_job = JobSubmission.objects.get(jobId=job_id)
+                    prev_job.delete()
```
is this from CEPv1.  i wonder why they would delete the previous job.  🤔   it would be fine for it to still exist, right?  not certain just thinking "out loud"

duckonomy replied:
@nathanfranklin Nope.. this was one of my questions.
I wanted to show that the same job (in the queue) was re-running.
Would it make more sense to keep the previous job around?

rstijerina replied:
Yeah, without good reasoning I think we should definitely preserve the previous job.

nathanfranklin replied:
> I wanted to show that the same job (in the queue) was re-running.

Because they would both have the old job and new job are going to have the same name, right?   yeah, i think i understand.  🤔  but then if we deleted the previous job, then users would no longer be able to get to the old jobs archived data (i.e View in Data Files button). I think 🤔

rstijerina commented:
> 1. Should we redesign it so that clicking on Relaunch Job opens another form that let's the user modify parameters based on the previous job submission?

Let's ask CMD, I'm not sure on the best use case scenario.

> 2. Would it be correct to show the Relaunch Job button when job status is failed as well as finished?

I think so, but we can confirm with CMD.

> 3. Should we implement cancelJob() when job status is not finished? (code was already there in the backend)

No, I don't think so.

> 4. I currently remove the previous job on clicking Relaunch Job and create a new job in it's place. Is this acceptable?

Let's remove that feature.

> 5. Should there be a separate RESUBMITTING job status?

I think the typical statuses will work ok

tacc-wbomar commented:
> > 1. Should we redesign it so that clicking on Relaunch Job opens another form that let's the user modify parameters based on the previous job submission?
> 
> Let's ask CMD, I'm not sure on the best use case scenario.

For _simulation_ jobs, **yes please**.
For _interactive_ app job, **no need**.

<details><summary>Source</summary>

Design (C.J.) in Slack: [Yes](https://tacc-team.slack.com/archives/CQUA4D5KJ/p1637610724113300?thread_ts=1637362349.106600&cid=CQUA4D5KJ), but [only for simulation job](https://tacc-team.slack.com/archives/CQUA4D5KJ/p1637612083114900?thread_ts=1637362349.106600&cid=CQUA4D5KJ).

</details>

> > 2. Would it be correct to show the Relaunch Job button when job status is failed as well as finished?
> 
> I think so, but we can confirm with CMD.

Yes, please. _Also, please change "Relaunch" to "Resubmit"._

<details><summary>Source</summary>

Design (C.J.) in Slack: [Yes __and__ change "Relaunch" to "Resubmit"](https://tacc-team.slack.com/archives/CQUA4D5KJ/p1637610724113300?thread_ts=1637362349.106600&cid=CQUA4D5KJ).

</details>

> > 5. Should there be a separate RESUBMITTING job status?
> 
> I think the typical statuses will work ok

Me too.

tacc-wbomar reviewed (changes requested):
UX flow—sans [expected change](https://github.com/TACC/Core-Portal/pull/552#issuecomment-981871444)[^1]—works. Notifications work. Two small UI comments.

[^1]: Will new UX (redirect to app form to change parameters) be for this ticket or a future ticket?

duckonomy commented:
@tacc-wbomar @xandroc @rstijerina 
### Changes
1. Changed 
style of button to look more like the [design docs](https://xd.adobe.com/view/db2660cc-1011-4f26-5d31-019ce87c1fe8-ad17/screen/7d3e671f-bb61-4a20-89c6-9ca9a78a6f93/). And changed the font-size to 16px.
2. Removed deleting previous job when resubmitting.
3. Redirect to `AppForm` when the app is a simulation app.

![Screenshot from 2021-12-02 12-06-45](https://user-images.githubusercontent.com/9425579/144478603-7b494acd-d645-4dbe-b373-919c825d1b70.png)

#### Potential problems
1. When redirecting to the app form for simulation applications, the application information is based on the `jobDetails` (as that the state we get from the `JobHistoryModal`). This does not seem perfectly correspond to the input values of each application. So, currently, I just copy over the `parameters` and `inputs` for the resubmitting application.
2. Similar to 1, I derive the original `appId` from the `jobDetails`, which looks something like `prtl.clone.USERNAME.ALLOCATION.-APPID-1.0`. I try to split and splice to get the value, which seems super hacky. Would there be a better way to do this? This is needed to go to the target application route.

tacc-wbomar commented on client/src/components/History/HistoryViews/JobHistoryModal.module.scss:
```diff
@@ -45,6 +45,11 @@
+.submit-button {
+    composes: c-button--primary from '../../../styles/components/c-button.css';
+    margin: auto;
+}
```
I think there is a misunderstanding.

Please change button text size _from_ 16px _to_ 14px.

<details><summary>Details</summary>

The font size was already 16px before c997bb0. That previous 16px was derived as [a computed value based on the `1rem` from `.btn` from Bootstrap styles](https://user-images.githubusercontent.com/62723358/143923659-108795c4-3fba-4b12-9583-a4ccd0065380.png).

<details><summary>More Details</summary>

The value `1rem` is computed as `html` font size x 1. The default `html` font-size for most browsers is 16px.

</details>

</details>

tacc-wbomar reviewed (commented):
> Redirect to AppForm when the app is a simulation app.

**Cool!** Confirmed for "Compress" and "Frontera HPC Jupyter" jobs.

> And changed the font-size to 16px.

Please see https://github.com/TACC/Core-Portal/pull/552#discussion_r761369682.

duckonomy commented:
> > Redirect to AppForm when the app is a simulation app.
> 
> **Cool!** Confirmed for "Compress" and "Frontera HPC Jupyter" jobs.
> 
> > And changed the font-size to 16px.
> 
> Please see [#552 (comment)](https://github.com/TACC/Core-Portal/pull/552#discussion_r761369682).

@tacc-wbomar Oops, I misread it :)

rstijerina commented on client/src/utils/jobsUtil.js:
```diff
@@ -128,6 +128,8 @@ export function getJobDisplayInformation(job, app) {
         display.nodeCount = job.nodeCount;
       }
+
+      display.interactive = app.definition.tags.includes('Interactive');
```
Maybe we should expand this to also include "VNC" and "DCV" tags - I can't say for sure if we add "Interactive" to all interactive apps

rstijerina commented on client/src/components/History/HistoryViews/JobHistoryModal.js:
```diff
@@ -98,6 +130,9 @@ function JobHistoryContent({ jobDetails, jobDisplay, jobName }) {
     configDataObj['Temporary Working Directory'] = jobDetails.workPath;
 
+  const jobFinished =
+    jobDetails.status === 'FINISHED' || jobDetails.status === 'FAILED';
```
We can use the `isTerminalState` function in `jobsUtil.js`, which also includes `STOPPED`

rstijerina commented:
> 1. When redirecting to the app form for simulation applications, the application information is based on the jobDetails (as that the state we get from the JobHistoryModal). This does not seem perfectly correspond to the input values of each application. So, currently, I just copy over the parameters and inputs for the resubmitting application.

This sounds ok, but can you give me an example of a discrepancy?

> 2. Similar to 1, I derive the original appId from the jobDetails, which looks something like prtl.clone.USERNAME.ALLOCATION.-APPID-1.0. I try to split and splice to get the value, which seems super hacky. Would there be a better way to do this? This is needed to go to the target application route.

Hacky indeed. I wonder if we can add some metadata to the job object with the original APPID, to make this easier. Similar to how we used pass through the Interactive Session Link in CEPv1 / DesignSafe. For now this seems like it will have to work. Perhaps we can create a util to convert between cloned app IDs and original app IDs

duckonomy commented:
> > 1. When redirecting to the app form for simulation applications, the application information is based on the jobDetails (as that the state we get from the JobHistoryModal). This does not seem perfectly correspond to the input values of each application. So, currently, I just copy over the parameters and inputs for the resubmitting application.
> 
> This sounds ok, but can you give me an example of a discrepancy?
> 

@rstijerina One would be the `processorsOnEachNode` on the app form, I think we could derive this from `processorsPerNode` value in the `jobDetails`. 
I'm not sure if they correspond directly to the similar sounding values and which parts we would have to manually translate to fill out the form from values from `jobDetails`.

tacc-wbomar reviewed (approved):
UI/UX approved.

rstijerina reviewed (approved):
LGTM. One question about the Resubmit Button - should the Job Detail modal close after the job resubmits?
